decisions/websre/0005-web-sre-kubernetes-secrets-pattern.md

```diff
@@ -0,0 +1,40 @@
+# 1. web-sre-kubernetes-secrets-pattern
+
+Date: 2021-06-09
+
+## Status
+
+Accepted
+
+## Context
+
+Secrets are inconsistently managed across our various clusters.  This makes it hard to trace where secrets come from and update them.  It also increases our exposure to security issues, as our each tool used can become compromised independently.  Current known secret management includes git-crypt repos, with sops, that apply kubernetes secrets direct to the cluster; sealed secrets; and external secrets pointing at cloud provider secret storage.  There are a few other methods that treat things that should be secrets incorrectly. For example, baking the 'secrets' into the docker image directly. We will treat those situations with more urgency, viewing them as explicitly a bad state to be in.
+
+The goal of this document is to describe the happy path for passing secrets to kubernetes based applications.
+
+## Decision
+
+Store the secrets in the cloud provider secret manager (aws secret manager, gcp secret manager), then use external secrets (https://github.com/external-secrets/kubernetes-external-secrets) to retrieve those secrets and present them to pods as env variables.  The reason this storage method is preferred is that the cloud secret managers seem to be the best secured, most tested, easiest to use place to store secrets. External secrets connects to the secret managers and presents the secret(s) in the method most often recommended for kubernetes; the 12 Factor App pattern of using environment variables.  One other nice benefit of this system is that references to the secrets can be stored in source code, so the connection between the app and the secret storage is explicit, but we're not checking the secret into git anywhere, risking leaking it. 
+
+There are currently a few ways the secrets can be written to the cloud secret store. We have some terraform that writes the secrets directly to the secrets stores, but most secrets are populated manually by operators.  We expect both methods to be used going forward.  In general, we expect SREs/Operators with elevated privileges to manage the secrets.  This ADR intentionally doesn't include scope for devs to write new secrets, or update existing ones.  That problem is as of today, unsolved.
```

Comment:
> That problem is as of today, unsolved.

I don't think this sentence should be included; it's redundant in light of the previous sentence, it doesn't contribute anything to this ADR, and it'll be invalidated once we make a decision and record it.